# Exercise Week 9 Association Rule Mining
Vania Vidya Vimala/0706022010013

## Arti Association Rules 

Association Rules adalah suatu metode data mining yang bertujuan untuk mencari sekumpulan items yang sering muncul bersamaan.
Association rule mining adalah salah satu metode data mining yang dapat mengidentifikasi hubungan kesamaan antar item. 
Dalam metode ini dibutuhkan algoritma untuk mencari kandidat aturan asosiasi. Salah satu algoritma yang biasa dipakai adalah algoritma apriori.

## Algoritma Apriori

Algoritma adalah urutan atau langkah-langkah yang logis dan sistematis untuk memecahkan dan menyelesaikan suatu masalah. Algoritma tidak selalu identik dengan ilmu komputer saja, dalam kehidupan sehari-hari pun banyak terdapat proses yang dinyatakan dalam suatu algoritma, dan algoritma pun juga merupakan kumpulan perintah untuk menyelesaikan suatu masalah.

Algoritma Apriori, Apriori yaitu termasuk jenis aturan asosiasi atau association rule dalam data mining. Aturan yang menyatakan asosiasi antara beberapa atribut sering disebut affinity analysis atau market basket analysis.

## Tipe Association rule 

1. Multi-relational association rules
2. Generalized Association rule
3. Interval Information Association Rules
4. Quantitative Association Rules

## Contoh Penerapan Association Rule

Menganalisis transaksi penjualan di suatu toko / warung, disetiap  pembelian yang dilakukan oleh konsumen, akan tercatat ID transaksi dan pada setiap transaksi dan memungkinkan ada satu atau lebih barang (item) yang dibeli, jika ada 100 transaksi setiap harinya, maka sudah ada 3000 transaksi dalam 1 bulan, nah dari transaksi inilah terbentuk pola-pola kombinasi pembelian item barang, seperti misal Konsumen A membeli  { beras, telur dan minyak }, Konsumen B membeli { telur dan minyak }  dan seterusnya.

Lalu peran Association Rule pada Algoritma Apriori disini diterapkan untuk membaca pola atau kecendrungan pembelian konsumen dengan melakukan perhitungan dari pembentukan suatu pola asosiasi produk pada data transaksi penjualan,  untuk menampilkan informasi tentang hasil informasi berupa pembentukan suatu pola asosiasi produk yang paling sering dibeli konsumen.

## Analisis Asosiasi dan Algoritma Apriori

Beberapa istilah yang sering digunakan pada Association Rule adalah:

- Support, adalah suatu ukuran yang menunjukkan seberapa besar tingkat dominasi suatu item atau item set dari keseluruhan transaksi.

- Confidence, adalah suatu ukuran yang menunjukkan seberapa sering aturan itu terbukti benar. 

- Lift Ratio, adalah rasio dari nilai pengamatan Support, diharapkan jika dua aturan itu independen

## Tahapan Proses Algoritma Apriori

In [10]:
# Menginstall Library
!pip install pandas
!pip install numpy
!pip install apyori

  Created wheel for apyori: filename=apyori-1.1.2-py3-none-any.whl size=5975 sha256=bd3d38568c8914afeef4ebbea94dada9e2eb63b5e9243ead8f5bbe0051e8c2d1
  Stored in directory: c:\users\user\appdata\local\pip\cache\wheels\1b\02\6c\a45230be8603bd95c0a51cd2b289aefdd860c1a100eab73661
Successfully built apyori


In [11]:
# Memanggil library yang dibutuhkan
import pandas as pd
import numpy as np
from apyori import apriori

In [24]:
# Load data 
df= 'https://raw.githubusercontent.com/vanyavidyavimala/dataset-apriori/main/dataset_2021-06-10.csv'
dt = pd.read_csv(df)
dt

Tanggal ID Transaksi  \
0    2021-04-02       PJ0002   
1    2021-04-09       PJ0003   
2    2021-04-09       PJ0004   
3    2021-04-09       PJ0004   
4    2021-04-09       PJ0006   
..          ...          ...   
97   2021-06-06     PJ260099   
98   2021-06-06    PJ3900100   
99   2021-06-07    PJ5200101   
100  2021-06-08    PJ5400102   
101  2021-06-09    PJ2700104   

                                                 Items  
0         Grape Candy Duren Simadu,Filter Regular 50gr  
1    3P ONAN 50gr Pink,3P ONAN 50gr Hitam,Surya JR ...  
2    3PUTRI Cap Nona 100gr Pink,3P Cap Nona 100gr P...  
3    3PUTRI Cap Nona 100gr Pink,3P Cap Nona 100gr P...  
4    3P ONAN 50gr Pink,3PUTRI Cap Nona 50gr merah,S...  
..                                                 ...  
97   3PUTRI Cap Nona 120gr Hitam,3P Cap Nona 120gr ...  
98   3PUTRI Cap Nona 100gr Pink,3P Cap Nona 100gr Pink  
99   3P ONAN 50gr Putih,Marlboro Merah JR Rokok 100...  
100  Marlboro Menthol JR Rokok 50gr,Warning 50gr,Bl...  
101  3PUTRI Cap Nona 120gr Hitam,3P Cap Nona 120gr ...  

[102 rows x 3 columns]

In [26]:
# Membuang kolom Tanggal, ID Transaksi
data=dt.drop(['Tanggal','ID Transaksi'],axis=1)

In [32]:
# Menggunakan fungsi apriori untuk membuat asosiasi
association_rules = apriori(trx, min_support=0.14, min_confidence=0.80,min_lift=1)
# Membuat list hasil dari algoritma apriori
association_results = association_rules

In [33]:
# Menampilkan hasil asosiasi dari item 
pd.set_option('max_colwidth', 1000)
Result=pd.DataFrame(columns=['Rule','Support','Confidence'])
for item in association_results:
    pair = item[2]
    for i in pair:
        items = str([x for x in i[0]])
        if i[3]!=1:
            Result=Result.append({
                'Rule':str([x for x in i[0]])+ " -> " +str([x for x in i[1]]),
                'Support':str(round(item[1]*100,2))+'%',
                'Confidence':str(round(i[2] *100,2))+'%'
                },ignore_index=True)
Result

Rule Support Confidence
0  ['Alat Linting Kecil Regular'] -> ['Buffalo Baret']  14.71%     83.33%
1        ['Lem V-tro kecil 10gr'] -> ['Buffalo Baret']  21.57%      88.0%

# Kesimpulan penerapan association rule diatas

Jika membeli Alat Linting Kecil Regular maka akan juga membeli Buffalo Baret dengan memiliki confidence atau tingkat kepercayaan sebesar 83.33%.

Jika membeli Lem V-tro kecil 10gr maka akan juga membeli Buffalo Baret dengan memiliki confidence atau tingkat kepercayaan sebesar 88.0%.

# Resources

- Agusta Y. (2008, Agustus 4). Association Rules.yudiagusta.wordpress.com.https://yudiagusta.wordpress.com/2008/08/04/association-rules/
- Desi U. (2018, Desember 21). Pengertian dan penerapan Metode Association Rules. garudacyber.co.id. https://garudacyber.co.id/artikel/1479-pengertian-dan-penerapan-metode-association-rules
- AnishaD. (2022, August 23). Association Rule. geeksforgeeks.org.https://www.geeksforgeeks.org/association-rule/
- Miftah S. (2021, Juli 10).Market Basket Analysis Menggunakan Machine Learning Python. dqlab.id.https://www.dqlab.id/market-basket-analysis-menggunakan-machine-learning-python
- Falah F. (2022, Juni 11). Penerapan Algoritma Apriori: Contoh dan Implementasi menggunakan python. Codekop.com.https://www.codekop.com/read/penerapan-algoritma-apriori-contoh-dan-implementasi-menggunakan-python-68.html
- Alfasaleh. (2019, September 19). Algoritma Apriori : Contoh Sederhana Penerapan Algoritma Apriori pada Association Rule. alfasoleh.com. https://www.alfasoleh.com/2019/09/data-mining-contoh-sederhana-penerapan.html
- Atmadjaya L. (2019). ASSOCIATION RULE DAN PENERAPANNYA. slideplayer.info. https://slideplayer.info/slide/16035867/
- Kusrini, &. E. (2009). Algoritma Data Mining. Yogyakarta: Andi Offset.
- Prasetyo, E. (2014). Data Mining. Yogyakarta: Andi Offset.
- Larose, D. T. (2014). Discovering Knowledge in Data: An Introduction to Data Mining Second Edition. New Jersey: John Wiley & Sons Inc